src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -257,7 +261,39 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 slice: None,
                 suffix: Vec::new(),
             },
-            _ => PatKind::Constant { value: cv },
+            _ => {
+                let mut leaf_ty = cv.ty;
+
+                // HACK(eddyb) workaround missing reference destructuring.
+                loop {
```

Comment:
I'm a little surprised to see there not a pre-existing method that does this traversal to find the content underlying a ref/array/slice? (Is the presupposition that we do not want such a method because any such traversal is almost certainly a hack, and should be annotated as such, as eddyb as done here?)

Reply:
This is that thing where I was *very* confused that this code doesn't handle references at all.
IIRC, @oli-obk's #70743 would make this unnecessary and we'd be correct by default.